common/frame_session.go

```diff
@@ -776,94 +781,109 @@ func (fs *FrameSession) onPageNavigatedWithinDocument(event *cdppage.EventNaviga
 }
 
 func (fs *FrameSession) onAttachedToTarget(event *target.EventAttachedToTarget) {
+	var (
+		ti  = event.TargetInfo
+		sid = event.SessionID
+		err error
+	)
+
 	fs.logger.Debugf("NewFrameSession:onAttachedToTarget",
 		"sid:%v tid:%v esid:%v etid:%v ebctxid:%v type:%q",
 		fs.session.id, fs.targetID, event.SessionID,
 		event.TargetInfo.TargetID, event.TargetInfo.BrowserContextID,
 		event.TargetInfo.Type)
 
 	session := fs.page.browserCtx.conn.getSession(event.SessionID)
-	targetID := event.TargetInfo.TargetID
-
-	if event.TargetInfo.Type == "iframe" {
-		frame := fs.manager.getFrameByID(cdp.FrameID(targetID))
-		if frame == nil {
-			return
-		}
-
-		fs.manager.removeChildFramesRecursively(frame)
-
-		// Successful case
-		frameSession, err := NewFrameSession(fs.ctx, session, fs.page, fs, targetID, fs.logger)
-		if err == nil {
-			fs.page.attachFrameSession(cdp.FrameID(targetID), frameSession)
-			return
-		}
-		// Erroneous cases
-		defer fs.logger.Debugf("FrameSession:onAttachedToTarget:NewFrameSession",
-			"sid:%v tid:%v esid:%v etid:%v ebctxid:%v type:%q err:%v",
-			fs.session.id, fs.targetID, event.SessionID,
-			event.TargetInfo.TargetID, event.TargetInfo.BrowserContextID,
-			event.TargetInfo.Type, err)
-		// ignore errors if we're no longer connected to browser
-		// this happens when there is no browser but we still want to
-		// attach a frame to it.
-		if !fs.page.browserCtx.browser.IsConnected() {
-			return
-		}
-		// Ignore context canceled error to gracefuly handle shutting down
-		// of the extension. This may happen because of generated events
-		// while a frame session is being created.
-		if errors.Is(err, context.Canceled) {
-			return
-		}
-		// Final chance:
-		// Throw a k6 error if the error is something different than a context
-		// cancelation.
-		select {
-		case <-fs.ctx.Done():
-			// ignore
-			return
-		default:
-			k6Throw(fs.ctx, "cannot create frame session (iframe): %w", err)
-		}
-	}
-
-	if event.TargetInfo.Type != "worker" {
-		fs.logger.Debugf("FrameSession:onAttachedToTarget returns: target is not a worker",
-			"sid:%v tid:%v esid:%v etid:%v ebctxid:%v type:%q",
+	if session == nil {
```

Comment:
Is this the fail-fast fix you mention? It's interesting that a session isn't available at this point considering that we always handle session creation/destruction before forwarding a CDP message to the higher layers (see https://github.com/grafana/xk6-browser/blob/8c7fc3b5e38e41488c833f38853190cd679c2589/common/connection.go#L280-L304).

Reply:
No, this is not the fail-fast fix :) I explained it in my other comment [here](https://github.com/grafana/xk6-browser/pull/169#issuecomment-1004843769).

This check is needed because sometimes there is a small window where a session can get removed by the connection while a new `FrameSession` is being constructed.

The `Session` is closed and removed by the `Connection`:

```
TRAC[0018] <- {"method":"Target.detachedFromTarget","params":{"sessionId":"CB2EC5F869C50B25582486DA64B58901","targetId":"CE598CB4776BC46D2EDFD02BE605CB4B"},"sessionId":"34759DABAC917ACD14805A6CCD588A99"}  category="cdp:recv" elapsed="0 ms" goroutine=40
DEBU[0018] sid:CB2EC5F869C50B25582486DA64B58901 tid:CE598CB4776BC46D2EDFD02BE605CB4B wsURL:ws://127.0.0.1:55791/devtools/browser/e5adcaf1-d538-4f52-bb1e-fa1f0f04eac9  category="Connection:closeSession" elapsed="0 ms" goroutine=40
DEBU[0018] sid:CB2EC5F869C50B25582486DA64B58901 tid:CE598CB4776BC46D2EDFD02BE605CB4B  category="Session:close" elapsed="0 ms" goroutine=40
DEBU[0018] sid:CB2EC5F869C50B25582486DA64B58901 tid:CE598CB4776BC46D2EDFD02BE605CB4B  category="Session:readLoop:<-s.done" elapsed="0 ms" goroutine=8742
```

Right after, a new `FrameSession` is getting created (`Session` becomes `nil` here):

```
DEBU[0018] sid:34759DABAC917ACD14805A6CCD588A99 tid:3A13FFBEA431F41D61631412B66A4563 esid:CB2EC5F869C50B25582486DA64B58901 etid:CE598CB4776BC46D2EDFD02BE605CB4B ebctxid:3A834A21E6848FC15E6107ADAF48043B type:"iframe"  category="FrameSession:onAttachedToTarget" elapsed="0 ms" goroutine=104
```

This happens 10% of the time among all the new `FrameSession` initializations.

Reply:
Ah I see, thanks for clarifying.